File: pyro_integration/workers/scanner.py
"""
R-YARA Scanner Worker

Worker implementation for YARA rule scanning tasks.
"""

import asyncio
import json
import subprocess
import tempfile
import os
from typing import Dict, Any, List, Optional
from pathlib import Path

from .base import RYaraWorker
from ..shared.protocol import WorkerTask, TaskResult, TaskType


class ScannerWorker(RYaraWorker):
    """
    Scanner worker for YARA rule matching.

    Handles:
    - File scanning with YARA rules
    - Memory/data scanning
    - Rule compilation and validation
    """

    @property
    def worker_type(self) -> str:
        return "r-yara-scanner"

    @property
    def capabilities(self) -> list:
        return [
            TaskType.SCAN_FILE.value,
            TaskType.SCAN_DATA.value,
            TaskType.VALIDATE_RULE.value,
            TaskType.COMPILE_RULES.value
        ]

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._yara_binary = self._find_yara_binary()
        self._yarac_binary = self._find_yarac_binary()

    def _find_yara_binary(self) -> Optional[str]:
        """Find YARA binary in PATH or known locations"""
        locations = [
            "yara",  # In PATH
            "/usr/bin/yara",
            "/usr/local/bin/yara",
            str(Path(__file__).parent.parent.parent / "yara"),
        ]
        for loc in locations:
            try:
                result = subprocess.run(
                    [loc, "--version"],
                    capture_output=True,
                    timeout=5
                )
                if result.returncode == 0:
                    return loc
            except (subprocess.SubprocessError, FileNotFoundError):
                continue
        return None

    def _find_yarac_binary(self) -> Optional[str]:
        """Find YARAC binary in PATH or known locations"""
        locations = [
            "yarac",  # In PATH
            "/usr/bin/yarac",
            "/usr/local/bin/yarac",
            str(Path(__file__).parent.parent.parent / "yarac"),
        ]
        for loc in locations:
            try:
                result = subprocess.run(
                    [loc, "--version"],
                    capture_output=True,
                    timeout=5
                )
                if result.returncode == 0:
                    return loc
            except (subprocess.SubprocessError, FileNotFoundError):
                continue
        return None

    async def process_task(self, task: WorkerTask) -> TaskResult:
        """Process a scanner task"""
        if task.task_type == TaskType.SCAN_FILE:
            return await self._scan_file(task)
        elif task.task_type == TaskType.SCAN_DATA:
            return await self._scan_data(task)
        elif task.task_type == TaskType.VALIDATE_RULE:
            return await self._validate_rule(task)
        elif task.task_type == TaskType.COMPILE_RULES:
            return await self._compile_rules(task)
        else:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error=f"Unknown task type: {task.task_type}"
            )

    async def _scan_file(self, task: WorkerTask) -> TaskResult:
        """Scan a file with YARA rules"""
        if not self._yara_binary:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error="YARA binary not found"
            )

        file_path = task.payload.get("file_path")
        rules = task.payload.get("rules")
        rules_file = task.payload.get("rules_file")

        if not file_path:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error="Missing file_path in payload"
            )

        if not rules and not rules_file:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error="Missing rules or rules_file in payload"
            )

        try:
            # Create temp file for rules if provided inline
            temp_rules_file = None
            if rules:
                temp_rules_file = tempfile.NamedTemporaryFile(
                    mode='w',
                    suffix='.yar',
                    delete=False
                )
                temp_rules_file.write(rules)
                temp_rules_file.close()
                rules_file = temp_rules_file.name

            # Run YARA scan
            cmd = [self._yara_binary, "-s", "-m", rules_file, file_path]

            process = await asyncio.create_subprocess_exec(
                *cmd,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE
            )

            stdout, stderr = await process.communicate()

            # Clean up temp file
            if temp_rules_file:
                os.unlink(temp_rules_file.name)

            # Parse results
            matches = self._parse_yara_output(stdout.decode())

            return TaskResult(
                task_id=task.task_id,
                success=True,
                data={
                    "matches": matches,
                    "match_count": len(matches),
                    "file_path": file_path
                }
            )

        except Exception as e:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error=str(e)
            )

    async def _scan_data(self, task: WorkerTask) -> TaskResult:
        """Scan raw data with YARA rules"""
        if not self._yara_binary:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error="YARA binary not found"
            )

        data = task.payload.get("data")
        rules = task.payload.get("rules")
        rules_file = task.payload.get("rules_file")

        if not data:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error="Missing data in payload"
            )

        if not rules and not rules_file:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error="Missing rules or rules_file in payload"
            )

        try:
            # Create temp file for data
            temp_data_file = tempfile.NamedTemporaryFile(
                mode='wb',
                delete=False
            )
            if isinstance(data, str):
                temp_data_file.write(data.encode())
            else:
                temp_data_file.write(data)
            temp_data_file.close()

            # Create temp file for rules if provided inline
            temp_rules_file = None
            if rules:
                temp_rules_file = tempfile.NamedTemporaryFile(
                    mode='w',
                    suffix='.yar',
                    delete=False
                )
                temp_rules_file.write(rules)
                temp_rules_file.close()
                rules_file = temp_rules_file.name

            # Run YARA scan
            cmd = [self._yara_binary, "-s", "-m", rules_file, temp_data_file.name]

            process = await asyncio.create_subprocess_exec(
                *cmd,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE
            )

            stdout, stderr = await process.communicate()

            # Clean up temp files
            os.unlink(temp_data_file.name)
            if temp_rules_file:
                os.unlink(temp_rules_file.name)

            # Parse results
            matches = self._parse_yara_output(stdout.decode())

            return TaskResult(
                task_id=task.task_id,
                success=True,
                data={
                    "matches": matches,
                    "match_count": len(matches)
                }
            )

        except Exception as e:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error=str(e)
            )

    async def _validate_rule(self, task: WorkerTask) -> TaskResult:
        """Validate YARA rule syntax"""
        if not self._yarac_binary:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error="YARAC binary not found"
            )

        rule_content = task.payload.get("rule")
        if not rule_content:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error="Missing rule in payload"
            )

        try:
            # Create temp files
            temp_rule_file = tempfile.NamedTemporaryFile(
                mode='w',
                suffix='.yar',
                delete=False
            )
            temp_rule_file.write(rule_content)
            temp_rule_file.close()

            temp_output = tempfile.NamedTemporaryFile(
                suffix='.yarc',
                delete=False
            )
            temp_output.close()

            # Try to compile
            cmd = [self._yarac_binary, temp_rule_file.name, temp_output.name]

            process = await asyncio.create_subprocess_exec(
                *cmd,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE
            )

            stdout, stderr = await process.communicate()

            # Clean up
            os.unlink(temp_rule_file.name)
            if os.path.exists(temp_output.name):
                os.unlink(temp_output.name)

            if process.returncode == 0:
                return TaskResult(
                    task_id=task.task_id,
                    success=True,
                    data={"valid": True, "message": "Rule is valid"}
                )
            else:
                return TaskResult(
                    task_id=task.task_id,
                    success=True,
                    data={
                        "valid": False,
                        "errors": stderr.decode().strip().split('\n')
                    }
                )

        except Exception as e:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error=str(e)
            )

    async def _compile_rules(self, task: WorkerTask) -> TaskResult:
        """Compile YARA rules to binary format"""
        if not self._yarac_binary:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error="YARAC binary not found"
            )

        rules = task.payload.get("rules")
        output_path = task.payload.get("output_path")

        if not rules:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error="Missing rules in payload"
            )

        try:
            # Create temp file for rules
            temp_rule_file = tempfile.NamedTemporaryFile(
                mode='w',
                suffix='.yar',
                delete=False
            )
            temp_rule_file.write(rules)
            temp_rule_file.close()

            # Use provided output path or temp file
            if not output_path:
                temp_output = tempfile.NamedTemporaryFile(
                    suffix='.yarc',
                    delete=False
                )
                temp_output.close()
                output_path = temp_output.name

            # Compile
            cmd = [self._yarac_binary, temp_rule_file.name, output_path]

            process = await asyncio.create_subprocess_exec(
                *cmd,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE
            )

            stdout, stderr = await process.communicate()

            # Clean up source
            os.unlink(temp_rule_file.name)

            if process.returncode == 0:
                # Read compiled rules if temp file
                compiled_data = None
                if task.payload.get("return_data", False):
                    with open(output_path, 'rb') as f:
                        import base64
                        compiled_data = base64.b64encode(f.read()).decode()

                return TaskResult(
                    task_id=task.task_id,
                    success=True,
                    data={
                        "compiled": True,
                        "output_path": output_path,
                        "compiled_data": compiled_data
                    }
                )
            else:
                return TaskResult(
                    task_id=task.task_id,
                    success=False,
                    error=stderr.decode().strip()
                )

        except Exception as e:
            return TaskResult(
                task_id=task.task_id,
                success=False,
                error=str(e)
            )

    def _parse_yara_output(self, output: str) -> List[Dict[str, Any]]:
        """Parse YARA scan output"""
        matches = []
        current_match = None

        for line in output.strip().split('\n'):
            if not line:
                continue

            # Match line format: rule_name [tag1,tag2] file_path
            if not line.startswith('0x'):
                parts = line.split()
                if len(parts) >= 2:
                    rule_name = parts[0]
                    # Extract tags if present
                    tags = []
                    if '[' in line and ']' in line:
                        tag_start = line.index('[') + 1
                        tag_end = line.index(']')
                        tags = [t.strip() for t in line[tag_start:tag_end].split(',')]

                    current_match = {
                        "rule": rule_name,
                        "tags": tags,
                        "strings": []
                    }
                    matches.append(current_match)
            else:
                # String match line format: 0xOFFSET:$identifier: content
                if current_match and ':' in line:
                    parts = line.split(':', 2)
                    if len(parts) >= 2:
                        current_match["strings"].append({
                            "offset": parts[0],
                            "identifier": parts[1],
                            "data": parts[2] if len(parts) > 2 else ""
                        })

        return matches


def create_scanner_worker(**kwargs) -> ScannerWorker:
    """Factory function to create a scanner worker"""
    return ScannerWorker(**kwargs)

File: rust/r-yara-pyro/src/workers/scanner.rs
//! Scanner worker implementation
//!
//! Handles YARA rule scanning, validation, and compilation tasks.
//! Now powered by r-yara-scanner for unified YARA scanning.

use std::path::PathBuf;
use std::time::Instant;
use tokio::process::Command;

use crate::protocol::{TaskResult, TaskType, WorkerTask};

use super::base::{BaseWorker, WorkerState, WorkerStats};

// Import r-yara-scanner
use r_yara_scanner::{Scanner, MetaValue};

/// Scanner worker for YARA operations
/// Now powered by r-yara-scanner - no external binaries needed!
pub struct ScannerWorker {
    base: BaseWorker,
    // Keep yarac for compile_rules (fallback for binary output)
    yarac_binary: Option<PathBuf>,
}

impl ScannerWorker {
    /// Create a new scanner worker
    pub fn new() -> Self {
        let base = BaseWorker::new(
            "r-yara-scanner",
            vec![
                TaskType::ScanFile,
                TaskType::ScanData,
                TaskType::ValidateRule,
                TaskType::CompileRules,
            ],
        );

        // Only look for yarac for compile_rules (binary output)
        let yarac_binary = Self::find_binary("yarac");

        Self {
            base,
            yarac_binary,
        }
    }

    /// Find a binary in PATH or known locations
    fn find_binary(name: &str) -> Option<PathBuf> {
        // Check if binary is in PATH
        if let Ok(output) = std::process::Command::new("which")
            .arg(name)
            .output()
        {
            if output.status.success() {
                let path = String::from_utf8_lossy(&output.stdout)
                    .trim()
                    .to_string();
                if !path.is_empty() {
                    return Some(PathBuf::from(path));
                }
            }
        }

        // Check known locations
        let locations = vec![
            format!("/usr/bin/{}", name),
            format!("/usr/local/bin/{}", name),
            format!("/opt/yara/bin/{}", name),
        ];

        for loc in locations {
            let path = PathBuf::from(&loc);
            if path.exists() {
                return Some(path);
            }
        }

        None
    }

    /// Get worker ID
    pub fn worker_id(&self) -> &str {
        &self.base.worker_id
    }

    /// Get worker type
    pub fn worker_type(&self) -> &str {
        &self.base.worker_type
    }

    /// Get capabilities
    pub fn capabilities(&self) -> Vec<TaskType> {
        self.base.capabilities.clone()
    }

    /// Get current state
    pub async fn state(&self) -> WorkerState {
        self.base.get_state().await
    }

    /// Get statistics
    pub async fn stats(&self) -> WorkerStats {
        self.base.get_stats().await
    }

    /// Process a task
    pub async fn process_task(&self, task: WorkerTask) -> TaskResult {
        let start = Instant::now();
        self.base.task_started(&task).await;

        let result = match task.task_type {
            TaskType::ScanFile => self.scan_file(&task).await,
            TaskType::ScanData => self.scan_data(&task).await,
            TaskType::ValidateRule => self.validate_rule(&task).await,
            TaskType::CompileRules => self.compile_rules(&task).await,
            _ => TaskResult::failure(&task.task_id, "Unsupported task type"),
        };

        let elapsed = start.elapsed().as_millis() as u64;
        self.base
            .task_completed(&task.task_id, result.success, elapsed)
            .await;

        result.with_execution_time(elapsed)
    }

    /// Scan a file with YARA rules using r-yara-scanner
    async fn scan_file(&self, task: &WorkerTask) -> TaskResult {
        let file_path = match task.payload.get("file_path") {
            Some(serde_json::Value::String(p)) => p,
            _ => return TaskResult::failure(&task.task_id, "Missing file_path in payload"),
        };

        let rules = task.payload.get("rules");
        let rules_file = task.payload.get("rules_file");

        // Get rules source
        let rules_source = match (rules, rules_file) {
            (Some(serde_json::Value::String(r)), _) => r.clone(),
            (_, Some(serde_json::Value::String(rf))) => {
                match tokio::fs::read_to_string(rf).await {
                    Ok(content) => content,
                    Err(e) => {
                        return TaskResult::failure(
                            &task.task_id,
                            format!("Failed to read rules file: {}", e),
                        );
                    }
                }
            }
            _ => return TaskResult::failure(&task.task_id, "Missing rules or rules_file in payload"),
        };

        // Create scanner
        let scanner = match Scanner::new(&rules_source) {
            Ok(s) => s,
            Err(e) => {
                return TaskResult::failure(
                    &task.task_id,
                    format!("Failed to compile rules: {}", e),
                );
            }
        };

        // Scan file
        match scanner.scan_file(file_path) {
            Ok(matches) => {
                let matches_json: Vec<serde_json::Value> = matches
                    .iter()
                    .map(|m| {
                        serde_json::json!({
                            "rule": m.rule_name.as_str(),
                            "tags": m.tags.iter().map(|t| t.as_str()).collect::<Vec<_>>(),
                            "strings": m.strings.iter().map(|s| {
                                serde_json::json!({
                                    "identifier": s.identifier.as_str(),
                                    "offsets": s.offsets
                                })
                            }).collect::<Vec<_>>(),
                            "meta": m.meta.iter().map(|(k, v)| {
                                let value = match v {
                                    MetaValue::String(s) => serde_json::Value::String(s.to_string()),
                                    MetaValue::Integer(i) => serde_json::Value::Number((*i).into()),
                                    MetaValue::Boolean(b) => serde_json::Value::Bool(*b),
                                    MetaValue::Float(f) => serde_json::json!(f),
                                };
                                (k.as_str(), value)
                            }).collect::<std::collections::HashMap<_, _>>()
                        })
                    })
                    .collect();

                TaskResult::success(
                    &task.task_id,
                    serde_json::json!({
                        "matches": matches_json,
                        "match_count": matches.len(),
                        "file_path": file_path
                    }),
                )
            }
            Err(e) => TaskResult::failure(&task.task_id, format!("Scan failed: {}", e)),
        }
    }

    /// Scan raw data with YARA rules using r-yara-scanner
    async fn scan_data(&self, task: &WorkerTask) -> TaskResult {
        let data = match task.payload.get("data") {
            Some(serde_json::Value::String(d)) => d,
            _ => return TaskResult::failure(&task.task_id, "Missing data in payload"),
        };

        let rules = task.payload.get("rules");
        let rules_file = task.payload.get("rules_file");

        // Get rules source
        let rules_source = match (rules, rules_file) {
            (Some(serde_json::Value::String(r)), _) => r.clone(),
            (_, Some(serde_json::Value::String(rf))) => {
                match tokio::fs::read_to_string(rf).await {
                    Ok(content) => content,
                    Err(e) => {
                        return TaskResult::failure(
                            &task.task_id,
                            format!("Failed to read rules file: {}", e),
                        );
                    }
                }
            }
            _ => return TaskResult::failure(&task.task_id, "Missing rules or rules_file in payload"),
        };

        // Create scanner
        let scanner = match Scanner::new(&rules_source) {
            Ok(s) => s,
            Err(e) => {
                return TaskResult::failure(
                    &task.task_id,
                    format!("Failed to compile rules: {}", e),
                );
            }
        };

        // Scan data
        match scanner.scan_bytes(data.as_bytes()) {
            Ok(matches) => {
                let matches_json: Vec<serde_json::Value> = matches
                    .iter()
                    .map(|m| {
                        serde_json::json!({
                            "rule": m.rule_name.as_str(),
                            "tags": m.tags.iter().map(|t| t.as_str()).collect::<Vec<_>>(),
                            "strings": m.strings.iter().map(|s| {
                                serde_json::json!({
                                    "identifier": s.identifier.as_str(),
                                    "offsets": s.offsets
                                })
                            }).collect::<Vec<_>>(),
                            "meta": m.meta.iter().map(|(k, v)| {
                                let value = match v {
                                    MetaValue::String(s) => serde_json::Value::String(s.to_string()),
                                    MetaValue::Integer(i) => serde_json::Value::Number((*i).into()),
                                    MetaValue::Boolean(b) => serde_json::Value::Bool(*b),
                                    MetaValue::Float(f) => serde_json::json!(f),
                                };
                                (k.as_str(), value)
                            }).collect::<std::collections::HashMap<_, _>>()
                        })
                    })
                    .collect();

                TaskResult::success(
                    &task.task_id,
                    serde_json::json!({
                        "matches": matches_json,
                        "match_count": matches.len()
                    }),
                )
            }
            Err(e) => TaskResult::failure(&task.task_id, format!("Scan failed: {}", e)),
        }
    }

    /// Validate YARA rule syntax using r-yara-scanner
    async fn validate_rule(&self, task: &WorkerTask) -> TaskResult {
        let rule = match task.payload.get("rule") {
            Some(serde_json::Value::String(r)) => r,
            _ => return TaskResult::failure(&task.task_id, "Missing rule in payload"),
        };

        // Try to compile the rule
        match Scanner::new(rule) {
            Ok(_) => TaskResult::success(
                &task.task_id,
                serde_json::json!({
                    "valid": true,
                    "message": "Rule is valid"
                }),
            ),
            Err(e) => TaskResult::success(
                &task.task_id,
                serde_json::json!({
                    "valid": false,
                    "errors": vec![e.to_string()]
                }),
            ),
        }
    }

    /// Compile YARA rules
    async fn compile_rules(&self, task: &WorkerTask) -> TaskResult {
        let yarac = match &self.yarac_binary {
            Some(path) => path,
            None => return TaskResult::failure(&task.task_id, "YARAC binary not found"),
        };

        let rules = match task.payload.get("rules") {
            Some(serde_json::Value::String(r)) => r,
            _ => return TaskResult::failure(&task.task_id, "Missing rules in payload"),
        };

        let output_path = task
            .payload
            .get("output_path")
            .and_then(|p| p.as_str())
            .map(PathBuf::from)
            .unwrap_or_else(|| {
                std::env::temp_dir().join(format!("yara_compiled_{}.yarc", task.task_id))
            });

        // Create temp file for rules
        let temp_rule_path =
            std::env::temp_dir().join(format!("yara_compile_{}.yar", task.task_id));

        if let Err(e) = tokio::fs::write(&temp_rule_path, rules).await {
            return TaskResult::failure(&task.task_id, format!("Failed to write rules: {}", e));
        }

        // Compile
        let temp_rule_str = temp_rule_path.to_string_lossy().to_string();
        let output_path_str = output_path.to_string_lossy().to_string();
        let output = Command::new(yarac)
            .arg(&temp_rule_str)
            .arg(&output_path_str)
            .output()
            .await;

        // Clean up source
        let _ = tokio::fs::remove_file(&temp_rule_path).await;

        match output {
            Ok(out) => {
                if out.status.success() {
                    // Optionally read compiled data
                    let compiled_data = if task
                        .payload
                        .get("return_data")
                        .and_then(|v| v.as_bool())
                        .unwrap_or(false)
                    {
                        match tokio::fs::read(&output_path).await {
                            Ok(data) => Some(base64::Engine::encode(
                                &base64::engine::general_purpose::STANDARD,
                                &data,
                            )),
                            Err(_) => None,
                        }
                    } else {
                        None
                    };

                    TaskResult::success(
                        &task.task_id,
                        serde_json::json!({
                            "compiled": true,
                            "output_path": output_path.to_string_lossy(),
                            "compiled_data": compiled_data
                        }),
                    )
                } else {
                    let stderr = String::from_utf8_lossy(&out.stderr);
                    TaskResult::failure(&task.task_id, stderr.to_string())
                }
            }
            Err(e) => TaskResult::failure(&task.task_id, format!("Compilation failed: {}", e)),
        }
    }

}

impl Default for ScannerWorker {
    fn default() -> Self {
        Self::new()
    }
}

// Add base64 to Cargo.toml dependencies

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_scanner_worker_creation() {
        let worker = ScannerWorker::new();
        assert_eq!(worker.worker_type(), "r-yara-scanner");
    }
}
